q2_composition/assets/run_ancombc.R

```diff
@@ -69,20 +69,48 @@ output_loaf         <- opt$output_loaf
 if (!file.exists(inp_abundances_path)) {
   errQuit("Input file path does not exist.")
 } else {
-  otu_file <- t(read.delim(inp_abundances_path, check.names = FALSE,
-                            row.names = 1))
+  # Base R
+  # otu_file <- t(read.delim(inp_abundances_path,
+  #                          check.names = TRUE, row.names = 1,
+  #                          colClasses = list(X = "character")))
+  # # The sample IDs are in the first column, but this column has no name.
+  # # The empty name is changed to X by check.names = T
+  # # Then, we say to treat column X (sample IDs) as strings
+
+  # Tidyverse
+  otu_file <- inp_abundances_path |>
+    # The sample IDs are in the first column, which we treat as characters.
+    # All other columns are feature frequencies as doubles
+    read_tsv(col_names = T, col_types = cols("character", .default = col_double())) |>
+    # The first column is called "...1" during import, then we convert it to row names.
+    column_to_rownames("...1") |>
+    t()
   }
 
 if (!file.exists(inp_metadata_path)) {
   errQuit("Metadata file path does not exist.")
 } else {
-  metadata_file <- read.delim(inp_metadata_path, check.names = FALSE,
-                              row.names = 1)
+  md_file <- read.delim(inp_metadata_path,
+                        check.names = FALSE, row.names = 1,
+                        colClasses = list(`sample-id` = "character"))
```

Comment:
@gregcaporaso 

This updated function supports sampleIDs like `1e002`

But it only worked by selecting a column called `sample-id`
So `sampleIDs ` and `sample-ids` would not work.

Can we fix this upstream with or do we have to support all valid first column names here?

Reply:
@colinbrislawn, I'm not certain. If the name of that column is normalized before it gets into R, then you should be set (and that would be the ideal way to handle it, so that we don't have to embed knowledge of the allowed column names here in addition to in the metadata parsing logic - i.e., the DRY principle). Have you tried passing the same metadata in to the QIIME 2 action with different column names? If you could try that and it works, you're all set and don't need to do it here. If it doesn't work, we should address that upstream (you may be able to rename that column before it's passed to R in the ANCOMBC action).

Reply:
Oh, that's a great idea.
I'll try this later today and report back!

Reply:
Okay,

This new function works for typical sample IDs and any `SampleID/sample id/#SampleID` column name.

However, it only can fix SI notation IDs (100e2) when the column is named `sample-id` due to how I identify this column.

>the name of that column is normalized before it gets into R

Yes, I'll see if I can get this added into this line:
https://github.com/qiime2/q2-composition/blob/dev/q2_composition/_ancombc.py#L89